### ISSUE TITLE
unsafe goal change

### ISSUE THREAD

**Issue by onlyjob** [lizardfs/lizardfs#331]
I have a goal "__std2__" defined as `$std {A B}` and a goal "__xor2__" defined as `$xor2 {A B C}`. Goals were specifically designed to tolerate failures of (less reliable) chunkservers in label C. 

I've changed goal from __std2__ to __xor2__ on bunch of files and shut down chunkserver C (before replication completed) resulting number of missing chunks in __xor2__.

This could only happen if one replica was deleted before chunk was converted to new goal __xor2__.

I _think_ currently goal changing is not safe: as soon as I've changed goal from __std2__ to __xor2__ many chunks were marked as _overgoal_ but it should have happened only after chunks conversion to __xor2__ to avoid losing replica. To avoid this scenario it is very important to ensure that cleanup recycles excess replicas only after migrating data to new goal, not before or during the process.

**Comment by DarkHaze** [comment]
I can guarantee that no chunk part (or subchunk) can be removed if this would make whole chunk not safe. Definition of safe state is that there are 2 whole copies of data available.
So if you have goal std2 being converted to xor2 then it's possible to remove one chunk part from std before we have all xor parts.
the state could look like 
std {A B} | xor2 {A C}
Now we can remove chunk part std {A} and we still have 2 full copies of data.
After this we have state:
std {A } | xor2 {A C}
Now it's not possible for any part to be removed because it would make chunk not safe.

Now about "overgoal" chunks. In code we have field that counts number of redundant chunk parts. After switching from std2 to xor2 you have exactly 2 chunk parts that should be removed (doesn't mean that can be removed at the moment because it could make chunk endangered). So the chunk is overgoaled.

To summarize: The chunk part can be removed only if the chunk after removing this part is safe. And safe chunk = chunk that have 2 full copies of data. I think that it is a reasonable approach.

I agree that cgi & stats need some change. With xors and standard chunk parts mixed, reporting number of chunk parts (as it was in 2.6) is now confusing and doesn't give full information about current status of chunks Some chunk parts correspond to whole data but some only to part of the data. If you count and show them in the same stats it's a little bit misleading. We are working at modifying cgi so it will be more readable with xors.

**Comment by onlyjob** [comment]
So you think it is merely a problem with CGI reporting instead of missing/unavailable __xor2__ chunk? That's great. I'm tempted to attempt reading all xor2 chunks while chunkserver C is down to ensure that there are no read errors. Unfortunately at the moment there are too many xor2 chunks to verify but I'll see if I can design an experiment to confirm that all "missing" block are actually readable...

**Comment by onlyjob** [comment]
I think I may not have isolated this problem well enough. _Some_ missing chunks may be attributed to prior chunkserver crashes #330 as every crash seems to cause a minor chunks loss (similar to unclean shutdown). However stopping chunkservers in label C reliably produces missing xor2 chunks in CGI... I'll wait till replication is completed and test further...

**Comment by onlyjob** [comment]
Maybe related to #252.

**Comment by onlyjob** [comment]
There is indeed a serious problem with goal change between _xor2_ and _std2_.
For example, I've changed goal of some files from _xor2_ to _std2_ and after some time (before goal change finished) I stopped one chunkserver. Immediately I got tens of thousands missing _std2_ chunks but that should have never happened if both chunks were created as new goal requires. There should have been undergoal/endangered chunks but not missing ones.
Presence of missing chunks while chunkserver is down demonstrate that LizardFS converted _xor2_ chunk into just one (degraded) replicated chunk (instead of two) hence I will lose my data if chunkserver (or its HDD) dies before replication create required number of replicas (as a matter of fact I've already lost more than 50 files due to chunkserver crashes #330 during goal change).

Goal change should not be dangerous. At the moment changing goal from _xor2_ to _std2_ is similar to recovering from loss of second replica in _std2_. Therefore changing goal is endangering data.

Goal change must be __transactional__. Master should only change type of chunk after  all replicas successfully created, as goal requires.

**Comment by psarna** [comment]
~~~
Presence of missing chunks while chunkserver is down demonstrate that LizardFS converted xor2 chunk into just one (degraded) replicated chunk
~~~
I dare say it's impossible to force master to delete a chunk copy if it would be in unsafe state after the operation. *Safe*, in this scenario, means *has 2 std copies*.

I'll try to reproduce this issue with stopping a chunkserver during replication, but I think the problem lies somewhere else.
Can you specify some steps to help me reproduce the issue?

One reasonable scenario is that some xor2 chunks were already undergoal (which may theoretically happen), like this:
1. xor2 chunk is undergoal:
`chunkserverA {  file1_xor2_part1  }; chunkserverB { file1_xor2_part2 }`
2. goal is changed to std and replication starts, creating an std copy on chunkserver A:
`chunkserverA {  file1_xor2_part1, file1_std_part  }; chunkserverB { file1_xor2_part2 }`
3. chunkserverA goes down:
 `chunkserverA { -  } , chunkserverB { file1_xor2_part2 }`
4. We've got a missing std2 chunk, but it was to be expected, since its xor2 parts were undergoal anyway.

Is it plausible in any way?

**Comment by onlyjob** [comment]
Thanks, Piotr. I hope this issue is pretty clear. I start with fully replicated cluster with no undergoal chunks, all chunkservers are up etc. Then I change goal of _xor2_ files to goal that defined as two replicas on labels 'A' and 'B'. In CGI graph "number of replications per minute" I see how chunks are copied to label 'B' first and only then to label 'A' while chunks appear as "undergoal" with only one replica in "Regular chunks state matrix". 
If I stop chunkserver before "undergoal" chunks are replicated I get unavailable files which is not only potentially dangerous for data safety but also bad for availability as goal 2 chunks should be available when one chunkserver is down. Here changing goal from _xor2_ to _2_ produces endangered chunks which should not be. I hope it is understandable...

**Comment by psarna** [comment]
The issue was here because it was possible, that the std copy was created on the same machine where one of xor copies lied. Then, if one of the xor copies was deleted (because std copy existed anyway), turning of chunkserver with both 1 xor copy and 1 std copy could result in missing chunk.

After the fix, the issue will only be here if there are exactly as many chunkservers in the system as required by xor level - e.g. 3 chunkservers total and xor2 goal present. This limitation must be here for the system to work smoothly and it will be added to documentation.

Long story short, after this fix is applied, please remember that if you use a xor goal,
it is better to have at least 1 more chunkserver in the whole cluster than the xor requires,
or changing a goal to std could temporarily endanger a chunk.

**Comment by onlyjob** [comment]
Thank you. I'll test that within next several hour and will let you know.

**Comment by onlyjob** [comment]
Sorry Piotr, but this patch does _not_ fix the issue.

_xor2_ is defined as `xor2: $xor2 {poolA poolB poolC}`;
_std2_ is defined as `2: $std {poolB poolC}`.
_poolB_ have two chunkservers; _poolC_ have four chunkservers.
Changing goal from _xor2_ to _std2_ exhibit the same problem -- endangered chunks (and missing chunks when I stop any chunkserver in _poolB_). However it might be that patch reduced number of missing chunks on chunkserver stop but I'm not sure...

I always re-check fixes for my bug reports so I can close bugs for you when they are really fixed if you wish.

**Comment by psarna** [comment]
Does this problem exist on newly created chunks or old ones? On existing one, the problem is still the old one - some chunkservers just happen to store 1 std copy and 1 xor copy of a chunk, so turning them off might result in missing chunks. The patch should work for new replications, for the old one, what needs to be done is to wait until std chunks are replicated to two copies.

**Comment by onlyjob** [comment]
I'm not sure that I understand your question. I've repeated experiment from scratch -- applied patch, restarted master, waited until there are no undergoal chunks left, set new goal (2) on bunch of _xor2_ files and observed the problem again (also stopped/started chunkserver to be sure)... I suppose chunks are _new_ since they only existed as fully replicated _xor2_ chunks across three labels before goal change.

**Comment by psarna** [comment]
Confirmed, there is still something wrong when labels are specified. Will try to fix.

**Comment by onlyjob** [comment]
Thank you. Piotr, maybe you could consider giving me more rights on this bug tracker so I could reopen or tag? I will not abuse my powers. ;)

**Comment by psarna** [comment]
Issue: if additional chunkservers are available, but they reached they replication limit, it was decided to create a std copy on the same chunkserver as xor copy, which created a potential danger.

http://cr.skytechnology.pl:8081/#/c/2289/ can you verify?

**Comment by psarna** [comment]
As for the rights, I'm not sure how, but I will try to do so.

**Comment by onlyjob** [comment]
Thank you. I'll test c2289 soon (I have to wait until current replication is finished)...

**Comment by onlyjob** [comment]
Is there anything can be done to make it work in situation when there is only one chunkserver in label? Now I'm looking into situation when _std2_ defined as `2: $std {poolA poolB}` and there is only one chunkserver in _poolA_. It is the same problem, after change of goal from _xor2_ (`$xor2 {poolA poolB poolC}`)...

**Comment by onlyjob** [comment]
And there seems to be a way to safely create all replicas -- since _poolB_ have more than one chunkserver, goal change can be performed safely if second (i.e. last) _std2_ replica is created on  _poolA_ by replacing old _xor2_ chunk. I'm looking forward to test another patch addressing this problem. :)

**Comment by onlyjob** [comment]
Few days later (the fallout): I've noticed the following error logged by chunkserver from _poolB_:

    replication error: Status 'No such chunk' sent by chunkserver (server 192.168.0.250:9422)

After restart of chunkserver 192.168.0.250 (_poolA_) I got 170+ missing chunks in goal _std2_ (poolA poolB). 

Before _poolA_ chunkserver was restarted, CGI showed no undergoal chunks and prior (_xor2_ to _std2_) goal change/replication finished 24+ hours ago.

What bothers me is that I recognise problems beyond merely placement of chunks in unsafe way on goal change. Somehow master misses both chunks and do not even raise warnings in CGI while chunkservers was suppose to accommodate data but silently fail...

I wish 3.9.2 was less of a disaster. :( 

This problem seems related to #342.

**Comment by psarna** [comment]
> Is there anything can be done to make it work when there is only one chunkserver in label? Now I'm  looking into situation when std2 defined as 2: $std {poolA poolB} and there is only one chunkserver in  poolA. It is the same problem, after change of goal from xor2 ($xor2 {poolA poolB poolC})...

This problem should have been covered by patch eb9436b2d928184f771ccc04e5b1b73ea918b377.
According to it, if it is possible to replicate to a server that does not have **any** chunk of a particular id, it will be done, even if it means ignoring the label constraints. In your case, as long as a *xor* part lies on chunkserver A, *std* part would not be created there, because only 1 chunkserver has this label. Instead, if there is more than 1 chunkserver with label B, *std* part will be created exactly there. I confirmed it several times on installations that matched your config (chunkservers: 1xA, 2xB, 1xC, goals as in your case).

The only edge case is when there are exactly N chunkservers on the installation and goal is changed from *xorN-1*. Then, of course, it is not possible to create an *std* copy on a chunkserver that does not store another part of the same chunk. This special case will be added to LizardFS asap, but it should not be the issue in case you described.

Can you confirm that an installation running newest commits from tree still has this problem?
If yes, it would be great if you shared steps you performed to achieve the unsafe state,
since I was unable to reproduce it, though I tried hard.

**Comment by onlyjob** [comment]
I can only tell that I had this patch applied since 3 days ago and the only step to reproduce is to change goal of some files from _xor2_ to _std2_. Last batch of files were converted to goal requiring two replicas on poolA and poolB where poolB have two chunkserver and poolA have one. I convert between 200 and 500 GiB and got about 170 corrupted files as result... :( Yes this problem is with newest commits as I was hoping they would address the problem that I originally reported...
How comes master is ignoring "No such chunk" messages?

**Comment by onlyjob** [comment]
Although I have both "keep parts of the same chunk on different servers" patches, my installation do not have `LIZ_CSTOMA_*` commits as well as few commits patching tests exclusively. Other than this it is almost identical to "master".

**Comment by psarna** [comment]
0c7c292b1ad0789aa17a2cbe9133049cdc3916bf is a third one addressing the potentially unsafe goal change

**Comment by psarna** [comment]
Oh you should definitely have `LIZ_CSTOMA_*` commits merged, they are the ones that fix issues with xor & std chunks residing on the same server from the chunkserver side.

**Comment by onlyjob** [comment]
Thanks, will push changes ASAP. :)

**Comment by psarna** [comment]
And we encourage to pull the lastest git head ref and building from it, so _all_ changes from the last couple of days are applied - they all contribute to fixing your issues. Picking only selected commits might not do the job.

**Comment by onlyjob** [comment]
OK. Perhaps those fixes could be accommodated to point release ASAP? 3.9.2 is unsafe and the longer those fixes are delayed the greater chances that someone would try 3.9.2 and walk away. If I were evaluating 3.9.2 a year ago I probably wouldn't be using LizardFS by now...

**Comment by Zorlin** [comment]
I can say that at least one deployment (mine!) is on hold till these issues
are resolved.

**Comment by onlyjob** [comment]
Very wise, Benjamin. ;)

**Comment by onlyjob** [comment]
> your config (chunkservers: 1xA, 2xB, 1xC, goals as in your case).

Actually I have 4xC. Chunkservers labelled as per their availability and reliability. A is most reliable UPS-powered chunkserver(s); B - enterprise grade hardware but no UPS; C - volatile storage on workstations.

It is definitely much better/safer with 3rd patch. CGI still show endangered chunks after goal change but chunks are not really endangered as stopping any chunkserver (especially in _poolA_ which is the only one) do not produce missing chunks. I also confirmed that with `mfsfileinfo`. With redundancy in place goal change appears to be safe now. Thank you.

I'm still getting "No such chunk" messages from chunkservers but restarting chunkservers no longer cause data loss. (Restart of _poolA_ chunkserver before latest patches killed more that 1000 files but with latest patches several restarts of chunkservers did not cause data loss as far as I can tell). Though #342 is still an issue...
